src/types/mappingproxy.rs

```diff
@@ -0,0 +1,563 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use super::PyMapping;
+use crate::err::PyResult;
+use crate::types::dict::PyDictItem;
+#[cfg(all(test, not(PyPy)))]
+use crate::types::dict::{PyDictItems, PyDictKeys, PyDictValues};
+use crate::types::{IntoPyDict, PyAny, PyDict, PyIterator, PyList, PySequence};
+use crate::PyObject;
+#[cfg(not(PyPy))]
+use crate::{ffi, AsPyPointer, PyErr, PyTryFrom, Python, ToPyObject};
+use std::os::raw::c_int;
+
+#[inline]
+#[allow(non_snake_case)]
+unsafe fn PyDictProxy_Check(object: *mut crate::ffi::PyObject) -> c_int {
+    ffi::PyObject_TypeCheck(object, ffi::addr_of_mut_shim!(ffi::PyDictProxy_Type))
+}
+
+/// Represents a Python `mappingproxy`.
+#[repr(transparent)]
+pub struct PyMappingProxy(PyAny);
+
+pyobject_native_type!(
+    PyMappingProxy,
+    ffi::PyDictProxyObject,
+    ffi::PyDictProxy_Type,
+    #checkfunction=PyDictProxy_Check
+);
+
+impl PyMappingProxy {
+    /// Creates a mappingproxy from an object.
+    pub fn new(py: Python<'_>, elements: impl IntoPyDict) -> PyResult<&'_ PyMappingProxy> {
+        unsafe {
+            let dict = elements.into_py_dict(py);
+            let proxy = ffi::PyDictProxy_New(dict.as_ptr());
+            py.from_owned_ptr_or_err::<PyMappingProxy>(proxy)
+        }
+    }
+
+    /// Creates a new mappingproxy from the sequence given.
+    ///
+    /// The sequence must consist of `(PyObject, PyObject)`.
+    ///
+    /// Returns an error on invalid input. In the case of key collisions,
+    /// this keeps the last entry seen.
+    #[cfg(not(PyPy))]
+    pub fn from_sequence(py: Python<'_>, seq: PyObject) -> PyResult<&PyMappingProxy> {
+        unsafe {
+            let dict = py.from_owned_ptr::<PyDict>(PyDict::from_sequence(py, seq)?.as_ptr());
+            py.from_owned_ptr_or_err(ffi::PyDictProxy_New(dict.as_ptr()))
+        }
+    }
```

Comment:
I guess this is just a direct copy of the `PyDict` equivalent? Depending on what we decide for `PyMappingProxy::new`, this may or may not be necessary.